### PR TITLE
Add MLX op handler for aten.isinf

### LINKED ISSUE #18922: Good First Issue: Add MLX Op Handler for `aten.isinf`
### 🚀 The feature, motivation and pitch

# Good First Issue: Add MLX Op Handler for `aten.isinf`

## Summary

Add support for `aten.isinf` in the MLX delegate. This op checks for infinite values element-wise and is needed for numerical stability checks and gradient clipping.

## Background

The MLX delegate converts PyTorch aten ops into MLX graph nodes during export. When an aten op has no handler, it falls back to CPU execution, breaking the GPU acceleration pipeline. Adding a handler lets the op run on the Metal GPU via MLX.

### Approach: Decomposed handler (preferred)

`aten.isinf` can be decomposed using comparison with infinity:

```python
# isinf(x) = abs(x) == inf
```

This uses existing `AbsNode` and `EqualNode` which are already supported.

## Steps

1. **Add handler in `backends/mlx/ops.py`**
   ```python
   @REGISTRY.register(target=[torch.ops.aten.isinf.default])
   def _isinf_handler(P: MLXProgramBuilder, n: Node) -> Slot:
       """Handle aten.isinf - check for infinite values element-wise.
       
       isinf(x) is equivalent to abs(x) == inf.
       """
       args = P.args(n)
       require_args(args, 1, 1, "aten.isinf")
       require_kwargs(P.kwargs(n), set(), "aten.isinf")
       x = args[0]
       
       # Create abs(x)
       _, abs_tmp = P.make_tmp_slot()
       P.emit(
           AbsNode(
               x=P.slot_to_tid(x),
               out=P.slot_to_tid(abs_tmp),
           )
       )
       
       # Create inf constant
       inf_slot = emit_lifted_constant(P, float('inf'), torch.float32)
       
       # Compare abs(x) == inf
       out = P.make_or_get_slot(n)
       P.emit(
           EqualNode(
               a=P.slot_to_tid(abs_tmp),
               b=P.slot_to_tid(inf_slot),
               out=P.slot_to_tid(out),
           )
       )
       return out
   ```

2. **Add test in `backends/mlx/test/test_ops.py`**
   
   Use the existing `_make_unary_op_test` infrastructure with a custom input function that includes inf values:

   ```python
   def _inf_input_fn():
       """Return a callable(shape, dtype) that generates inputs with some inf values."""
       def fn(shape, dtype):
           x = torch.randn(shape, dtype=dtype)
           # Insert some inf values
           mask_pos = torch.rand(shape) > 0.8
           mask_neg = torch.rand(shape) > 0.9
           x[mask_pos] = float('inf')
           x[mask_neg] = float('-inf')
           return (x,)
       return fn
   
   # Add to _UNARY_OP_TESTS list:
   {"op_name": "isinf", "op_fn": torch.isinf, "shapes": _SHAPES_3, "input_fn": _inf_input_fn()},
   ```

### Running tests

```bash
python -m executorch.backends.mlx.test.run_all_tests -k isinf
```

## References

- MLX C++: `array isinf(const array &a, StreamOrDevice s = {})`
- PyTorch signature: `isinf(Tensor self) -> Tensor`
- Note: This catches both positive and negative infinity
- Test infrastructure: See `_make_unary_op_test` and `_UNARY_OP_TESTS` in `test_ops.py`

### Alternatives

_No response_

### Additional context

_No response_

### RFC (Optional)

_No response_

### PR DESCRIPTION
**Pull request by Ai-chan-0411**
## Summary

While looking at the MLX backend coverage for numerical-stability ops, I noticed `aten.isinf` was missing — any model using `torch.isinf` would silently fall back to CPU, breaking the GPU acceleration pipeline.

This PR adds a decomposed handler that expresses `isinf(x)` as `abs(x) == inf`, reusing the existing `AbsNode` and `EqualNode` infrastructure. Both positive and negative infinity are correctly detected through the abs step.

**Changes:**
- `backends/mlx/ops.py` — new `_isinf_handler` registered for `torch.ops.aten.isinf.default`
- `backends/mlx/test/test_ops.py` — added `_inf_input_fn` (generates tensors with scattered ±inf values) and an `isinf` entry in `_UNARY_OP_TESTS`

Closes #18922

cc @metascroy

### CODE REVIEW

File: backends/mlx/test/test_ops.py [hunk 1]
```diff
@@ -4004,6 +4004,20 @@ def fn(shape, dtype):
     return fn
 
 
+def _inf_input_fn():
+    """Return a callable(shape, dtype) that generates inputs with some inf values."""
+
+    def fn(shape, dtype):
+        x = torch.randn(shape, dtype=dtype)
```
Comment (metascroy) [inline backends/mlx/test/test_ops.py:4011]: Can we add some nans to this generated test input as well?

File: backends/mlx/ops.py [hunk 2]
```diff
@@ -2697,6 +2697,41 @@ def _relu_handler(P: MLXProgramBuilder, n: Node) -> Slot:
     return out
 
 
+@REGISTRY.register(target=[torch.ops.aten.isinf.default])
```
Comment (metascroy) [inline backends/mlx/ops.py:2700]: Are there any other variants of this torch op? E.g., isinf.Tensor, etc.

There may not be, just curious.

**Comment by metascroy** [comment]
Looks great @Ai-chan-0411! Can we add nans to the test case (to make sure isinf behaves correctly when its input is nan)?

**Comment by Ai-chan-0411** [comment]
Thanks for the review! You're right about adding NaN test cases. Since `isinf` returns False for NaNs (they're neither inf nor -inf), we should include them in the generated test inputs to ensure that edge case is properly covered. Let me update `_inf_input_fn` to mix in some NaN values alongside the current ±inf cases.

**Comment by Ai-chan-0411** [comment]
Good question about variants. In PyTorch's aten ops, `aten.isinf` is the primary op. The `.default` suffix we register for is the standard overload. There's no separate `isinf.Tensor` variant — torch.isinf() works on both Tensors and scalars through operator overloading at the higher level, but at the aten level it maps to the single aten.isinf.default op. I'll verify this is the only variant we need to handle.

File: backends/mlx/test/test_ops.py (hunk 1 above)
Comment (Ai-chan-0411) [reply backends/mlx/test/test_ops.py:4011]: Done — added NaN values to `_inf_input_fn`. The generated inputs now include `float("nan")` alongside `inf`/`-inf`, so the test verifies that `isinf` correctly returns `False` for NaN. Pushed in 7647a7e.

File: backends/mlx/ops.py (hunk 2 above)
Comment (Ai-chan-0411) [reply backends/mlx/ops.py:2700]: I checked the PyTorch ATen operator registry and the executorch codebase — the only variant is `aten.isinf.default`. There is no `isinf.Tensor` or `isinf.out` overload defined. The other backends (MPS, Qualcomm) also only register `aten.isinf.default`, so this handler covers the complete surface.

**Comment by Ai-chan-0411** [comment]
I noticed the MLX workflow run (24492915727) failed, but the failure was due to infrastructure issues — specifically, network connectivity errors when cloning submodules (NVIDIA/cutlass and google/flatbuffers). This is not related to the code changes in this PR. The implementation itself should be sound. A retry of the workflow run should succeed.

**Comment by metascroy** [comment]
Re-running workflows

**Review by metascroy — APPROVED** [review]

**Comment by metascroy** [comment]
@Ai-chan-0411 can you sign our Contributor License Agreement? See the failing "Meta CLA Check".

**Comment by metascroy** [comment]
@Ai-chan-0411 you can ignore the failing mlx-llm tests. I think there is an infra issue with running them on PRs from the community. The other MLX tests are passing.

Can you resolve the merge conflict, and we can merge?

**Comment by nil-is-all** [comment]
@Ai-chan-0411, could you sign the CLA and resolve merge conflicts to move forward?

**Comment by nil-is-all** [comment]
CLosing since #18988 addressed this
